packages/pkl.lua/lua.pkl

```diff
@@ -294,3 +332,560 @@ class Renderer extends ValueRenderer {
+  let (loc = errorLocation(source, token))
+  let (errMsg = "\(msg)\n\n\(loc.line)\n\(loc.marker)\nat \(uri ?? "<input>"):\(loc.row1):\(loc.col1)")
+    throw(errMsg)
+
+local const function throwError2(msg: String, source: String, uri: Uri?, token1: RegexMatch, note1: String, token2: RegexMatch, note2: String): nothing =
+  let (loc1 = errorLocation(source, token1))
+  let (loc2 = errorLocation(source, token2))
+  let (errMsg =
+    if (loc1.row1 == loc2.row1) // same line
+      if (loc1.col1 <= loc2.col1) // loc1 comes first
+        "\(msg)\n\n\(loc1.line)\n\(loc1.marker) \(note1)\n\(loc2.marker) \(note2)\nat \(uri ?? "<input>"):\(loc1.row1):\(loc1.col1)"
+      else // loc2 comes first
+        "\(msg)\n\n\(loc1.line)\n\(loc2.marker) \(note2)\n\(loc1.marker) \(note1)\nat \(uri ?? "<input>"):\(loc1.row1):\(loc2.col1)"
+    else if (loc1.row1 < loc2.row1) // loc1 comes first
+      "\(msg)\n\n\(loc1.line)\n\(loc1.marker) \(note1)\n\(loc2.line)\n\(loc2.marker) \(note2)\nat \(uri ?? "<input>"):\(loc1.row1):\(loc1.col1)"
+    else // loc2 comes first
+      "\(msg)\n\n\(loc2.line)\n\(loc2.marker) \(note2)\n\(loc1.line)\n\(loc1.marker) \(note1)\nat \(uri ?? "<input>"):\(loc2.row1):\(loc2.col1)")
+    throw(errMsg)
+
+local const function throwExpected(expected: String, source: String, uri: Uri?, token: RegexMatch?): nothing =
+  let (found =
+    if (token == null) "EOF"
+    else if (token.value == "\n") "newline"
+    else if (token.value.isEmpty) "EOF"
+    else "token `\(token.value.replaceAllMapped(Regex(#"[\n\\]"#), (it) -> if (it.value == "\n") "\\n" else #"\\"#))`")
+    throwError("expected \(expected), found \(found)", source, uri, token ?? new RegexMatch { value = ""; start = source.length; end = source.length })
+
+// endregion
+// endregion
+// region Hex
+
+/// parseHex tranforms a single hexadecimal character into its unsigned integer representation.
+local const function parseHex(digit: Char): UInt8 = nybbleLut.getOrNull(digit) ?? throw("Unrecognized hex digit: \(digit)")
+
+/// parseHexOctet tranforms a two hexadecimal characters into its unsigned integer representation.
+local const function parseHexOctet(octet: String(length == 2)): UInt8 = byteLut.getOrNull(octet) ?? throw("Unrecognized hex octet: \(octet)")
+
+/// parseHex32 transforms an 8 character hexidecimal string into its UInt32 representation.
+local const function parseHex32(s: String(length == 8)): UInt32 =
+  IntSeq(0, 7)
+    .step(2)
+    .map((it) -> s.substring(it, it + 2))
+    .fold(0, (acc, it) -> acc.shl(8) + parseHexOctet(it))
+
+local const nybbleLut = new Mapping {
+  for (n in IntSeq(0, 9)) {
+    [n.toString()] = n
+  }
+  for (n in IntSeq(0xa, 0xf)) {
+    [n.toRadixString(16)] = n
+    [n.toRadixString(16).toUpperCase()] = n
+  }
+}
+
+local const byteLut = new Mapping {
+  for (k,v in nybbleLut) {
+    for (k2,v2 in nybbleLut) {
+      ["\(k)\(k2)"] = v.shl(4) + v2
+    }
+  }
+}
+
+// endregion
```

Comment:
These functions were based on some functions found in [pkl.experimental.net/net.pkl](https://github.com/apple/pkl-pantry/blob/437359005c0c4d2afdac2e2fb191740d3dc0a073/packages/pkl.experimental.net/net.pkl#L498-L515). The changes made here compared to the originals were done for performance reasons, as benchmarking showed that a lookup table is significantly faster than pkl.experimental.net's approach for `parseHex()`, and expanding the table to include uppercase letters is faster than lowercasing the input.

I am tempted to propose that we make a new package `pkl.hex` that contains these functions along with any other useful hex manipulation utilities, although it would probably be better just to update the Pkl stdlib to support parsing with radix natively.

Reply:
Yeah, we already have `Int#toRadixString`. I think it makes sense to have `String#toRadixInt()` as well, which also complements `String#toInt()`.